src/backend/utils/time/snapmgr.c

```diff
@@ -2245,6 +2250,47 @@ RestoreTransactionSnapshot(Snapshot snapshot, void *source_pgproc)
 
 /*
  * XidInMVCCSnapshot
+ *
+ * Check whether this xid is in snapshot. When enable_csn_snapshot is
+ * switched off just call XidInLocalMVCCSnapshot().
+ */
+bool
+XidInMVCCSnapshot(TransactionId xid, Snapshot snapshot)
+{
+	bool in_snapshot;
+
+	in_snapshot = XidInLocalMVCCSnapshot(xid, snapshot);
+
+	if (!get_csnlog_status())
+	{
+		Assert(XidCSNIsFrozen(snapshot->snapshot_csn));
+		return in_snapshot;
+	}
+
+	if (in_snapshot)
+	{
+		/*
+		 * This xid may be already in unknown state and in that case
+		 * we must wait and recheck.
+		 */
+		return XidInvisibleInCSNSnapshot(xid, snapshot);
+	}
+	else
+	{
+#ifdef USE_ASSERT_CHECKING
+		/* Check that csn snapshot gives the same results as local one */
+		if (XidInvisibleInCSNSnapshot(xid, snapshot))
+		{
+			XidCSN gcsn = TransactionIdGetXidCSN(xid);
+			Assert(XidCSNIsAborted(gcsn) || XidCSNIsInProgress(gcsn));
```

Comment:
Can the `XidCSNIsAborted(gcsn)` case ever happen? I don't think you ever need to check for the visibility of an aborted transaction.